lib/services/MetricService.ts

```diff
@@ -0,0 +1,206 @@
+ *
+ * Copyright 2015-2021 Kuzzle
+ * mailto: support AT kuzzle.io
+ * website: http://kuzzle.io
+ *
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { JSONObject } from 'kuzzle';
+import { Gauge, Registry, collectDefaultMetrics, Histogram } from 'prom-client';
+import { PrometheusPluginConfiguration } from '../PrometheusPlugin';
+
+/**
+ * Core metrics type definition
+ */
+export type CoreMetrics = {
+  api: {
+    /**
+     * Concurrent requests being processed
+     */
+    concurrentRequests: Gauge<string>;
+
+    /**
+     * Number of requests waiting to be processed
+     */
+    pendingRequests: Gauge<string>;
+  },
+  network: {
+    /**
+     * Number of active connections per protocol using labels
+     */
+    connections: Gauge<string>;
+  },
+  realtime: {
+    /**
+     * Number of active realtime rooms
+     * @see https://docs.kuzzle.io/core/2/guides/main-concepts/realtime-engine
+     */
+    rooms: Gauge<string>;
+
+    /**
+     * Number of active realtime subscriptions
+     * @see https://docs.kuzzle.io/core/2/guides/main-concepts/realtime-engine
+     */
+    subscriptions: Gauge<string>
+  }
+};
+
+
+/**
+ * MetricService is a service to handle metrics from the Kuzzle API and the application.
+ * @property {{ core: CoreMetrics, requestDuration?: Histogram<string> }} metrics     - The application metrics to register.
+ * @property {{[key: string]: Registry}}                                  registries  - The Prometheus registry to register metrics.
+ */
+export class MetricService {
+  /**
+   * A set of metrics matching the Kuzzle server:metrics action reponse
+   * @property {CoreMetrics} core            - The core metrics to register.
+   * @property {Histogram}   requestDuration - The application metrics to register.
+   */
+  private metrics: { core: CoreMetrics, requestDuration?: Histogram<string> };
+
+  /**
+   * The Prometheus registries used to register metrics and format them
+   * @property {Registry} core              - The core metrics registry
+   * @property {Registry} default           - The default metrics registry (if enabled)
+   * @property {Registry} requestDuration   - The default metrics registry (if enabled)
+   */
+  private registries: { core: Registry, default?: Registry, requestDuration?: Registry };
+
+  /**
+   * @param {PrometheusPluginConfiguration} config - The plugin configuration
+   */
+  constructor(config: PrometheusPluginConfiguration) {
```

Comment:
Nitpicking: coding standard
```suggestion
  constructor (config: PrometheusPluginConfiguration) {
```